include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,275 @@
+
+			for (uint16_t w = 0u; w < copyLayerCount; ++w)
+			{
+				std::array<decodeType, maxChannels> maxAbsoluteDecodeValues = {};
+
+				{
+					const uint8_t* inData = reinterpret_cast<const uint8_t*>(state->inImage->getBuffer()->getPointer());
+					const auto blockDims = asset::getBlockDimensions(state->inImage->getCreationParameters().format);
+					static constexpr uint8_t maxPlanes = 4;
+
+					auto decode = [&](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+					{
+						core::vectorSIMDu32 localOutPos = readBlockPos * blockDims - core::vectorSIMDu32(state->inOffset.x, state->inOffset.y, state->inOffset.z);
+
+						auto* inDataAdress = inData + readBlockArrayOffset;
+						const void* inSourcePixels[maxPlanes] = { inDataAdress, nullptr, nullptr, nullptr };
+
+						decodeType decodeBuffer[maxChannels] = {};
+						for (auto blockY = 0u; blockY < blockDims.y; blockY++)
+							for (auto blockX = 0u; blockX < blockDims.x; blockX++)
+							{
+								asset::decodePixelsRuntime(inFormat, inSourcePixels, decodeBuffer, blockX, blockY);
+								const size_t offset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(localOutPos.x + blockX, localOutPos.y + blockY, localOutPos.z), scratchByteStrides);
+								memcpy(reinterpret_cast<uint8_t*>(scratchMemory) + offset, decodeBuffer, scratchTexelByteSize);
+							}
+					};
+
+					IImage::SSubresourceLayers subresource = { static_cast<IImage::E_ASPECT_FLAGS>(0u), state->inMipLevel, state->inBaseLayer, 1 };
+					CMatchedSizeInOutImageFilterCommon::state_type::TexelRange range = { state->inOffset,state->extent };
+					CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, inFormat);
+
+					auto& inRegions = state->inImage->getRegions(state->inMipLevel);
+					CBasicImageFilterCommon::executePerRegion(state->inImage, decode, inRegions.begin(), inRegions.end(), clipFunctor);
+				}
+
+				{
+					auto getScratchPixel = [&](core::vector4di32_SIMD readBlockPos) -> decodeType*
+					{
+						const size_t scratchOffset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(readBlockPos.x, readBlockPos.y, readBlockPos.z, 0), scratchByteStrides);
+						return reinterpret_cast<decodeType*>(reinterpret_cast<uint8_t*>(scratchMemory) + scratchOffset);
+					};
+
+					auto computeDerivativeTexel = [&](core::vectorSIMDi32 readBlockPos) -> void
+					{
+						decodeType* current = getScratchPixel(readBlockPos);
+						auto& [x, y, z, a] = std::make_tuple(*current, *(current + 1), *(current + 2), *(current + 3));
+
+						std::for_each(current, current + currentChannelCount,
+							[&](const decodeType& itrValue)
+							{
+								uint8_t offset = &itrValue - current;
+								const decodeType absoluteValue = core::abs(itrValue);
+							
+								if (maxAbsoluteDecodeValues[offset] < absoluteValue)
+									maxAbsoluteDecodeValues[offset] = absoluteValue;
+							}
+						);
+
+						x = -x / z;
+						y = -y / z;
```

Comment:
nope, first do the divide, then take the max!

Also only consider the x and y channels for maxing

P.S. Move the divide by `z` up to the decode, then scratch only has to store `x` and `y` only

Reply:
corrected in following commit